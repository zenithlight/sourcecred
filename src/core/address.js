// @flow

import stringify from "json-stable-stringify";
import deepFreeze from "deep-freeze";

import * as MapUtil from "../util/map";
import * as C from "../util/combo";

export interface AddressModule<Address: string> {
  /**
   * Assert at runtime that the provided address is actually a valid
   * address of this kind, throwing an error if it is not. If `what` is
   * provided, it will be included in the error message.
   */
  assertValid(address: Address, what?: string): void;

  /**
   * Assert at runtime that the provided array is a valid array of
   * address parts (i.e., a valid input to `fromParts`), throwing an
   * error if it is not. If `what` is provided, it will be included in
   * the error message.
   */
  assertValidParts(parts: $ReadOnlyArray<string>, what?: string): void;

  /**
   * The empty address (the identity for `append`). Equivalent to
   * `fromParts([])`.
   */
  empty: Address;

  /**
   * Convert an array of address parts to an address. The input must be
   * a non-null array of non-null strings, none of which contains the
   * NUL character. This is the inverse of `toParts`.
   */
  fromParts(parts: $ReadOnlyArray<string>): Address;

  /**
   * Convert an address to the array of parts that it represents. This
   * is the inverse of `fromParts`.
   */
  toParts(address: Address): string[];

  /**
   * Pretty-print an address. The result will be human-readable and
   * contain only printable characters. Clients should not make any
   * assumptions about the format.
   */
  toString(address: Address): string;

  /**
   * Construct an address by extending the given address with the given
   * additional components. This function is equivalent to:
   *
   *     return fromParts([...toParts(address), ...components]);
   *
   * but may be more efficient.
   */
  append(address: Address, ...components: string[]): Address;

  /**
   * Test whether the given address has the given prefix. This function
   * is equivalent to:
   *
   *     const prefixParts = toParts(prefix);
   *     const addressParts = toParts(address);
   *     const actualPrefix = addressParts.slice(0, prefixParts.length);
   *     return deepEqual(prefix, actualPrefix);
   *
   * (where `deepEqual` checks value equality on arrays of strings), but
   * may be more efficient.
   *
   * Note that this is an array-wise prefix, not a string-wise-prefix:
   * e.g., `toParts(["ban"])` is not a prefix of `toParts(["banana"])`.
   */
  hasPrefix(address: Address, prefix: Address): boolean;

  /**
   * Interpret the provided string as an Address.
   *
   * Addresses are natively stored as strings. This method verifies
   * that the provided "raw" address is actually an Address, so that
   * you can have a type-level assurance that a string is an Address.
   *
   * This is useful if e.g. you are loading serialized Addresses.
   *
   * Throws an error if the string is not a valid Address.
   */
  fromRaw(raw: string): Address;

  /**
   * A parser for Addresses.
   *
   * Convenience wrapper `fromRaw`.
   */
  parser: C.Parser<Address>;
}

export type Options = {|
  /**
   * The name of this kind of address, like `NodeAddress`.
   */
  +name: string,

  /**
   * A unique nonce for the runtime representation of this address. For
   * compact serialization, this should be short; a single letter
   * suffices.
   */
  +nonce: string,

  /**
   * For the purposes of nice error messages: in response to an address
   * of the wrong kind, we can inform the user what kind of address they
   * passed (e.g., "expected NodeAddress, got EdgeAddress"). This
   * dictionary maps another address module's nonce to the name of that
   * module.
   */
  +otherNonces?: Map<string, string>,
|};

export function makeAddressModule(options: Options): AddressModule<string> {
  type Address = string; // for readability and interface consistency
  const {name, nonce} = options;
  const otherNonces = new Map(options.otherNonces || new Map());

  const separator = "\0";
  if (nonce.indexOf(separator) !== -1) {
    throw new Error(`invalid nonce (contains NUL): ${stringify(nonce)}`);
  }

  const nonceWithSeparator = nonce + separator;
  const otherNoncesWithSeparators = MapUtil.mapKeys(
    otherNonces,
    (otherNonce) => {
      if (otherNonce === nonce) {
        throw new Error(
          `primary nonce listed as otherNonce: ${stringify(nonce)}`
        );
      }
      if (otherNonce.indexOf(separator) !== -1) {
        throw new Error(
          `invalid otherNonce (contains NUL): ${stringify(otherNonce)}`
        );
      }
      return otherNonce + separator;
    }
  );

  function assertValid(address: Address, what?: string): void {
    // istanbul ignore if
    if (process.env.NODE_ENV !== "test") {
      // Catching invalid addresses in test code should be sufficient
      return;
    }
    const prefix = what == null ? "" : `${what}: `;
    if (address == null) {
      throw new Error(prefix + `expected ${name}, got: ${String(address)}`);
    }
    if (!address.endsWith(separator)) {
      throw new Error(prefix + `expected ${name}, got: ${stringify(address)}`);
    }
    if (!address.startsWith(nonceWithSeparator)) {
      for (const [
        otherNonceWithSeparator,
        otherName,
      ] of otherNoncesWithSeparators) {
        if (address.startsWith(otherNonceWithSeparator)) {
          throw new Error(
            prefix + `expected ${name}, got ${otherName}: ${stringify(address)}`
          );
        }
      }
      throw new Error(prefix + `expected ${name}, got: ${stringify(address)}`);
    }
  }

  function partsString(parts: $ReadOnlyArray<string>): string {
    // This is needed to properly print arrays containing `undefined`.
    return "[" + parts.map((p) => String(stringify(p))).join(",") + "]";
  }

  function assertValidParts(
    parts: $ReadOnlyArray<string>,
    what?: string
  ): void {
    // istanbul ignore if
    if (process.env.NODE_ENV !== "test") {
      // Catching invalid parts in test code should be sufficient
      return;
    }
    const prefix = what == null ? "" : `${what}: `;
    if (parts == null) {
      throw new Error(
        prefix + `expected array of parts, got: ${String(parts)}`
      );
    }
    parts.forEach((s: string) => {
      if (s == null) {
        throw new Error(
          prefix +
            `expected array of parts, got ${String(s)} in: ${partsString(
              parts
            )}`
        );
      }
      if (s.indexOf(separator) !== -1) {
        const where = `${stringify(s)} in ${partsString(parts)}`;
        throw new Error(prefix + `part contains NUL character: ${where}`);
      }
    });
  }

  function nullDelimited(components: $ReadOnlyArray<string>): string {
    return [...components, ""].join(separator);
  }

  function fromParts(parts: $ReadOnlyArray<string>): Address {
    assertValidParts(parts);
    return nonce + separator + nullDelimited(parts);
  }

  const empty = fromParts([]);

  function toParts(address: Address): string[] {
    assertValid(address);
    const parts = address.split(separator);
    return parts.slice(1, parts.length - 1);
  }

  function toString(address: Address): string {
    const parts = toParts(address);
    return `${name}${stringify(parts)}`;
  }

  function append(address: Address, ...parts: string[]): Address {
    assertValid(address);
    assertValidParts(parts);
    return address + nullDelimited(parts);
  }

  function hasPrefix(address: Address, prefix: Address): boolean {
    assertValid(address, "address");
    assertValid(prefix, "prefix");
    return address.startsWith(prefix);
  }

  function fromRaw(address: string): Address {
    if (!address.endsWith(separator)) {
      throw new Error(
        `address does not end with separator: ${stringify(address)}`
      );
    }
    if (!address.startsWith(nonceWithSeparator)) {
      for (const [
        otherNonceWithSeparator,
        otherName,
      ] of otherNoncesWithSeparators) {
        if (address.startsWith(otherNonceWithSeparator)) {
          throw new Error(
            `expected ${name}, got ${otherName}: ${stringify(address)}`
          );
        }
      }
      throw new Error(`expected ${name}, got: ${stringify(address)}`);
    }
    return address;
  }

  const parser: C.Parser<Address> = C.fmap(C.string, fromRaw);

  const result = {
    assertValid,
    assertValidParts,
    empty,
    fromParts,
    toParts,
    toString,
    append,
    hasPrefix,
    fromRaw,
    parser,
  };
  return deepFreeze(result);
}
