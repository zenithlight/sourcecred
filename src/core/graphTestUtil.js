// @flow

import deepFreeze from "deep-freeze";
import {
  EdgeAddress,
  Graph,
  NodeAddress,
  type Node,
  type Edge,
  type NodeAddressT,
  type EdgeAddressT,
} from "./graph";
import type {TimestampMs} from "../util/timestamp";

/**
 * Create a new Node from an array of string address parts.
 *
 * Fields on the node will be set with dummy values.
 * Test code should endeavor to use this whenever the code
 * is not testing how specific fields are handled; that way, adding
 * new fields will not require updating unrelated test code across
 * the codebase.
 *
 * The returned node is frozen; as such, it is safe to re-use this exact
 * object across test cases.
 */
export function partsNode(parts: string[]): Node {
  return deepFreeze({
    address: NodeAddress.fromParts(parts),
    description: parts.toString(),
    timestampMs: null,
  });
}

/**
 * Create a new Node from a single address part.
 *
 * The same considerations as partsNode apply.
 */
export function node(name: string): Node {
  return partsNode([name]);
}

/**
 * Create a new Edge from address parts and a src and dst.
 *
 * This is a convenience method for constructing example edges more concisely in test code.
 *
 * The returned edge is frozen, so it is safe to use across test cases.
 */
export function partsEdge(parts: string[], src: Node, dst: Node): Edge {
  return deepFreeze({
    address: EdgeAddress.fromParts(parts),
    src: src.address,
    dst: dst.address,
    timestampMs: 0,
  });
}

/**
 * Create a new Edge from a single address part and a src and dst.
 *
 * The same considerations as partsEdge apply.
 */
export function edge(name: string, src: Node, dst: Node): Edge {
  return partsEdge([name], src, dst);
}

export function advancedGraph(): {|
  edges: {|
    fullDanglingEdge: {|
      +address: EdgeAddressT,
      +dst: NodeAddressT,
      +src: NodeAddressT,
      +timestampMs: TimestampMs,
    |},
    halfDanglingEdge: {|
      +address: EdgeAddressT,
      +dst: NodeAddressT,
      +src: NodeAddressT,
      +timestampMs: TimestampMs,
    |},
    hom1: {|
      +address: EdgeAddressT,
      +dst: NodeAddressT,
      +src: NodeAddressT,
      +timestampMs: TimestampMs,
    |},
    hom2: {|
      +address: EdgeAddressT,
      +dst: NodeAddressT,
      +src: NodeAddressT,
      +timestampMs: TimestampMs,
    |},
    loopLoop: {|
      +address: EdgeAddressT,
      +dst: NodeAddressT,
      +src: NodeAddressT,
      +timestampMs: TimestampMs,
    |},
    phantomEdge1: {|
      +address: EdgeAddressT,
      +dst: NodeAddressT,
      +src: NodeAddressT,
      +timestampMs: TimestampMs,
    |},
    phantomEdge2: {|
      +address: EdgeAddressT,
      +dst: NodeAddressT,
      +src: NodeAddressT,
      +timestampMs: TimestampMs,
    |},
  |},
  graph1: () => Graph,
  graph2: () => Graph,
  nodes: {|
    dst: {|
      +address: NodeAddressT,
      +description: string,
      +timestampMs: TimestampMs | null,
    |},
    halfIsolated: {|
      +address: NodeAddressT,
      +description: string,
      +timestampMs: TimestampMs | null,
    |},
    isolated: {|
      +address: NodeAddressT,
      +description: string,
      +timestampMs: TimestampMs | null,
    |},
    loop: {|
      +address: NodeAddressT,
      +description: string,
      +timestampMs: TimestampMs | null,
    |},
    phantomNode: {|
      +address: NodeAddressT,
      +description: string,
      +timestampMs: TimestampMs | null,
    |},
    src: {|
      +address: NodeAddressT,
      +description: string,
      +timestampMs: TimestampMs | null,
    |},
  |},
|} {
  // The advanced graph has the following features:
  // - Multiple edges of same hom, from `src` to `dst`
  // - An isolated node, `isolated`
  // - A loop
  // - A node and edge with the same toParts representation
  // This function exposes all of the pieces of the advanced graph.
  // It also returns two different versions of the graph, which are
  // logically equivalent but very different history
  // To avoid contamination, every piece is exposed as a function
  // which generates a clean copy of that piece.
  const src = node("src");
  const dst = node("dst");
  const hom1 = partsEdge(["hom", "1"], src, dst);
  const hom2 = partsEdge(["hom", "2"], src, dst);
  const loop = node("loop");
  const loopLoop = edge("loop", loop, loop);

  const halfIsolated = node("halfIsolated");
  const phantomNode = node("phantom");
  const halfDanglingEdge = edge("half-dangling", halfIsolated, phantomNode);
  const fullDanglingEdge = edge("full-dangling", phantomNode, phantomNode);

  const isolated = node("isolated");
  const graph1 = () =>
    new Graph()
      .addNode(src)
      .addNode(dst)
      .addNode(loop)
      .addNode(isolated)
      .addEdge(hom1)
      .addEdge(hom2)
      .addEdge(loopLoop)
      .addNode(halfIsolated)
      .addEdge(halfDanglingEdge)
      .addEdge(fullDanglingEdge);

  // graph2 is logically equivalent to graph1, but is constructed with very
  // different history.
  // Use this to check that logically equivalent graphs are treated
  // equivalently, regardless of their history.
  const phantomEdge1 = edge("phantom", src, phantomNode);
  const phantomEdge2 = edge("not-so-isolated", src, isolated);

  // To verify that the graphs are equivalent, every mutation is preceded
  // by a comment stating what the set of nodes and edges are prior to that mutation
  const graph2 = () =>
    new Graph()
      // N: [], E: []
      .addNode(phantomNode)
      // N: [phantomNode], E: []
      .addNode(src)
      // N: [phantomNode, src], E: []
      .addEdge(phantomEdge1)
      // N: [phantomNode, src], E: [phantomEdge1]
      .addNode(isolated)
      // N: [phantomNode, src, isolated], E: [phantomEdge1]
      .addNode(halfIsolated)
      // N: [phantomNode, src, isolated, halfIsolated]
      // E: [phantomEdge1]
      .addEdge(halfDanglingEdge)
      // N: [phantomNode, src, isolated, halfIsolated]
      // E: [phantomEdge1, halfDanglingEdge]
      .addEdge(fullDanglingEdge)
      // N: [phantomNode, src, isolated, halfIsolated]
      // E: [phantomEdge1, halfDanglingEdge, fullDanglingEdge]
      .removeEdge(phantomEdge1.address)
      // N: [phantomNode, src, isolated, halfIsolated]
      // E: [halfDanglingEdge, fullDanglingEdge]
      .addNode(dst)
      // N: [phantomNode, src, isolated, halfIsolated, dst]
      // E: [halfDanglingEdge, fullDanglingEdge]
      .addEdge(hom1)
      // N: [phantomNode, src, isolated, halfIsolated, dst]
      // E: [halfDanglingEdge, fullDanglingEdge, hom1]
      .addEdge(phantomEdge2)
      // N: [phantomNode, src, isolated, halfIsolated, dst]
      // E: [halfDanglingEdge, fullDanglingEdge, hom1, phantomEdge2]
      .addEdge(hom2)
      // N: [phantomNode, src, isolated, halfIsolated, dst]
      // E: [halfDanglingEdge, fullDanglingEdge, hom1, phantomEdge2, hom2]
      .removeEdge(hom1.address)
      // N: [phantomNode, src, isolated, halfIsolated, dst]
      // E: [halfDanglingEdge, fullDanglingEdge, phantomEdge2, hom2]
      .removeNode(phantomNode.address)
      // N: [src, isolated, halfIsolated, dst]
      // E: [halfDanglingEdge, fullDanglingEdge, phantomEdge2, hom2]
      .removeEdge(phantomEdge2.address)
      // N: [src, isolated, halfIsolated, dst]
      // E: [halfDanglingEdge, fullDanglingEdge, hom2]
      .removeNode(isolated.address)
      // N: [src, halfIsolated, dst]
      // E: [halfDanglingEdge, fullDanglingEdge, hom2]
      .addNode(isolated)
      // N: [src, halfIsolated, dst, isolated]
      // E: [halfDanglingEdge, fullDanglingEdge, hom2]
      .addNode(loop)
      // N: [src, halfIsolated, dst, isolated, loop]
      // E: [halfDanglingEdge, fullDanglingEdge, hom2]
      .addEdge(loopLoop)
      // N: [src, halfIsolated, dst, isolated, loop]
      // E: [halfDanglingEdge, fullDanglingEdge, hom2, loopLoop]
      .addEdge(hom1);
  //     N: [src, halfIsolated, dst, isolated, loop]
  //     E: [halfDanglingEdge, fullDanglingEdge, hom2, loopLoop, hom1]
  const nodes = {src, dst, loop, isolated, phantomNode, halfIsolated};
  const edges = {
    hom1,
    hom2,
    loopLoop,
    phantomEdge1,
    phantomEdge2,
    halfDanglingEdge,
    fullDanglingEdge,
  };
  return {nodes, edges, graph1, graph2};
}
