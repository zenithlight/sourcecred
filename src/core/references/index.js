//@flow

export type {URL, ReferenceDetector} from "./referenceDetector";
export {MappedReferenceDetector} from "./mappedReferenceDetector";
export {CascadingReferenceDetector} from "./cascadingReferenceDetector";
export {TranslatingReferenceDetector} from "./translatingReferenceDetector";
