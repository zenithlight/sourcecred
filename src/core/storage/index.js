// @flow

/**
 * Data Storage allows the implementation of a uniform abstraction for I/O
 */
export interface DataStorage {
  /**
   * If the value for the key entered cannot be found, an error should be thrown.
   */
  get(key: string): Promise<Uint8Array>;
}

/**
 * keys should be file-system friendly
 */
export interface WritableDataStorage extends DataStorage {
  set(key: string, value: Uint8Array): Promise<void>;
}
