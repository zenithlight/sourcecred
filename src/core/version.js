// @flow

const packageJson = require("../../package.json");

export const VERSION_SHORT: string = `v${packageJson.version}`;
