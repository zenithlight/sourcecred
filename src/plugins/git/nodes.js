// @flow

import deepFreeze from "deep-freeze";
import {NodeAddress, type NodeAddressT} from "../../core/graph";
import type {Hash} from "./types";

export opaque type RawAddress: NodeAddressT = NodeAddressT;

const GIT_PREFIX = NodeAddress.fromParts(["sourcecred", "git"]);
export function _gitAddress(...parts: string[]): RawAddress {
  return NodeAddress.append(GIT_PREFIX, ...parts);
}

export const COMMIT_TYPE: "COMMIT" = "COMMIT";

export const Prefix: {|base: NodeAddressT, commit: RawAddress|} = deepFreeze({
  base: GIT_PREFIX,
  commit: _gitAddress(COMMIT_TYPE),
});

export type CommitAddress = {|
  +type: typeof COMMIT_TYPE,
  +hash: Hash,
|};

export type StructuredAddress = CommitAddress;

export function fromRaw(x: RawAddress): StructuredAddress {
  function fail() {
    return new Error(`Bad address: ${NodeAddress.toString(x)}`);
  }
  if (!NodeAddress.hasPrefix(x, GIT_PREFIX)) {
    throw fail();
  }
  const [_unused_sc, _unused_git, _type, ...rest] = NodeAddress.toParts(x);
  const type: $ElementType<StructuredAddress, "type"> = (_type: any);
  switch (type) {
    case "COMMIT": {
      if (rest.length !== 1) throw fail();
      const [hash] = rest;
      return {type: COMMIT_TYPE, hash};
    }
    default:
      // eslint-disable-next-line no-unused-expressions
      (type: empty);
      throw fail();
  }
}

export function toRaw(x: StructuredAddress): RawAddress {
  switch (x.type) {
    case COMMIT_TYPE:
      return NodeAddress.append(Prefix.commit, x.hash);
    default:
      throw new Error(`Unexpected type ${(x.type: empty)}`);
  }
}
