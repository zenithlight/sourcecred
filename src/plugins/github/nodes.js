// @flow

import deepFreeze from "deep-freeze";
import {NodeAddress, type NodeAddressT} from "../../core/graph";
import {botSet} from "./bots";

export opaque type RawAddress: NodeAddressT = NodeAddressT;

const GITHUB_PREFIX = NodeAddress.fromParts(["sourcecred", "github"]);
export function _githubAddress(...parts: string[]): RawAddress {
  return NodeAddress.append(GITHUB_PREFIX, ...parts);
}

export const REPO_TYPE: "REPO" = "REPO";
export const ISSUE_TYPE: "ISSUE" = "ISSUE";
export const PULL_TYPE: "PULL" = "PULL";
export const REVIEW_TYPE: "REVIEW" = "REVIEW";
export const COMMENT_TYPE: "COMMENT" = "COMMENT";
export const COMMIT_TYPE: "COMMIT" = "COMMIT";
export const USERLIKE_TYPE: "USERLIKE" = "USERLIKE";
export const USER_SUBTYPE: "USER" = "USER";
export const BOT_SUBTYPE: "BOT" = "BOT";

export const Prefix: {|
  base: NodeAddressT,
  bot: RawAddress,
  comment: RawAddress,
  commit: RawAddress,
  issue: RawAddress,
  issueComment: RawAddress,
  pull: RawAddress,
  pullComment: RawAddress,
  repo: RawAddress,
  review: RawAddress,
  reviewComment: RawAddress,
  user: RawAddress,
  userlike: RawAddress,
|} = deepFreeze({
  base: GITHUB_PREFIX,
  repo: _githubAddress(REPO_TYPE),
  issue: _githubAddress(ISSUE_TYPE),
  pull: _githubAddress(PULL_TYPE),
  review: _githubAddress(REVIEW_TYPE),
  comment: _githubAddress(COMMENT_TYPE),
  commit: _githubAddress(COMMIT_TYPE),
  userlike: _githubAddress(USERLIKE_TYPE),
  user: _githubAddress(USERLIKE_TYPE, USER_SUBTYPE),
  bot: _githubAddress(USERLIKE_TYPE, BOT_SUBTYPE),
  reviewComment: _githubAddress(COMMENT_TYPE, REVIEW_TYPE),
  issueComment: _githubAddress(COMMENT_TYPE, ISSUE_TYPE),
  pullComment: _githubAddress(COMMENT_TYPE, PULL_TYPE),
});

/**
 * Return the address corresponding to a GitHub login.
 *
 * If the login is considered a bot, then a bot address is returned. Otherwise,
 * a regular user address is returned. The method does not attempt to find out
 * whether the address should actually be an organization address, as we don't
 * yet handle organization addresses.
 *
 * Note: The signature will need to be refactored when we make the list of bots
 * a configuration option rather than a hardcoded constant.
 */
export function loginAddress(username: string): RawAddress {
  const bots = botSet();
  if (bots.has(username)) {
    return NodeAddress.append(Prefix.bot, username);
  } else {
    return NodeAddress.append(Prefix.user, username);
  }
}

export type RepoAddress = {|
  +type: typeof REPO_TYPE,
  +owner: string,
  +name: string,
|};
export type IssueAddress = {|
  +type: typeof ISSUE_TYPE,
  +repo: RepoAddress,
  +number: string,
|};
export type PullAddress = {|
  +type: typeof PULL_TYPE,
  +repo: RepoAddress,
  +number: string,
|};
export type ReviewAddress = {|
  +type: typeof REVIEW_TYPE,
  +pull: PullAddress,
  +id: string,
|};
export type CommentAddress = {|
  +type: typeof COMMENT_TYPE,
  +parent: CommentableAddress,
  +id: string,
|};
export type CommitAddress = {|
  +type: typeof COMMIT_TYPE,
  +id: string,
|};
export type UserlikeAddress = {|
  +type: typeof USERLIKE_TYPE,
  +subtype: typeof USER_SUBTYPE | typeof BOT_SUBTYPE,
  +login: string,
|};

export type StructuredAddress =
  | RepoAddress
  | IssueAddress
  | PullAddress
  | ReviewAddress
  | CommentAddress
  | CommitAddress
  | UserlikeAddress;

// Each of these types has 0 or more "AUTHORS" edges, each of which
// leads to a UserlikeAddress.  Note: It is not true that every
// Authorable has at least one author, as when GitHub accounts are
// deleted, they leave behind posts without authors.
export type AuthorableAddress =
  | IssueAddress
  | PullAddress
  | ReviewAddress
  | CommentAddress
  | CommitAddress;

// Each of these types has text content, which means
// it may be the source of a reference to a ReferentAddress.
export type TextContentAddress =
  | IssueAddress
  | PullAddress
  | ReviewAddress
  | CommentAddress
  | CommitAddress;

// Each of these types may be referred to by something
// with text content.
export type ReferentAddress =
  | RepoAddress
  | IssueAddress
  | PullAddress
  | ReviewAddress
  | CommentAddress
  | CommitAddress
  | UserlikeAddress;

// Each of these types is structurally the child of some
// entity with a ParentAddress.
export type ChildAddress =
  | IssueAddress // child of RepoAddress
  | PullAddress // child of RepoAddress
  | ReviewAddress // child of PullAddress
  | CommentAddress; // child of IssueAddress, PullAddress, or ReviewAddress

// Each of these types may have child nodes (see ChildAddress).
export type ParentAddress =
  | RepoAddress
  | IssueAddress
  | PullAddress
  | ReviewAddress;

// GitHub allows you to react to these types
export type ReactableAddress = IssueAddress | PullAddress | CommentAddress;

// Each of these types may have Comments as children
export type CommentableAddress = IssueAddress | PullAddress | ReviewAddress;

// Verify that Commentable is a subtype of Parent
const _unused_static = (_: CommentableAddress): ParentAddress => _;

export function fromRaw(x: RawAddress): StructuredAddress {
  function fail() {
    return new Error(`Bad address: ${NodeAddress.toString(x)}`);
  }
  if (!NodeAddress.hasPrefix(x, GITHUB_PREFIX)) {
    throw fail();
  }
  const [_unused_sc, _unused_gh, kind, ...rest] = NodeAddress.toParts(x);
  switch (kind) {
    case REPO_TYPE: {
      if (rest.length !== 2) {
        throw fail();
      }
      const [owner, name] = rest;
      return {type: REPO_TYPE, owner, name};
    }
    case ISSUE_TYPE: {
      if (rest.length !== 3) {
        throw fail();
      }
      const [owner, name, number] = rest;
      const repo = {type: REPO_TYPE, owner, name};
      return {type: ISSUE_TYPE, repo, number};
    }
    case PULL_TYPE: {
      if (rest.length !== 3) {
        throw fail();
      }
      const [owner, name, number] = rest;
      const repo = {type: REPO_TYPE, owner, name};
      return {type: PULL_TYPE, repo, number};
    }
    case REVIEW_TYPE: {
      if (rest.length !== 4) {
        throw fail();
      }
      const [owner, name, pullNumber, id] = rest;
      const repo = {type: REPO_TYPE, owner, name};
      const pull = {type: PULL_TYPE, repo, number: pullNumber};
      return {type: REVIEW_TYPE, pull, id};
    }
    case COMMENT_TYPE: {
      if (rest.length < 1) {
        throw fail();
      }
      const [subkind, ...subrest] = rest;
      switch (subkind) {
        case ISSUE_TYPE: {
          if (subrest.length !== 4) {
            throw fail();
          }
          const [owner, name, issueNumber, id] = subrest;
          const repo = {type: REPO_TYPE, owner, name};
          const issue = {type: ISSUE_TYPE, repo, number: issueNumber};
          return {type: COMMENT_TYPE, parent: issue, id};
        }
        case PULL_TYPE: {
          if (subrest.length !== 4) {
            throw fail();
          }
          const [owner, name, pullNumber, id] = subrest;
          const repo = {type: REPO_TYPE, owner, name};
          const pull = {type: PULL_TYPE, repo, number: pullNumber};
          return {type: COMMENT_TYPE, parent: pull, id};
        }
        case REVIEW_TYPE: {
          if (subrest.length !== 5) {
            throw fail();
          }
          const [owner, name, pullNumber, reviewFragment, id] = subrest;
          const repo = {type: REPO_TYPE, owner, name};
          const pull = {type: PULL_TYPE, repo, number: pullNumber};
          const review = {type: REVIEW_TYPE, pull, id: reviewFragment};
          return {type: COMMENT_TYPE, parent: review, id};
        }
        default:
          throw fail();
      }
    }
    case COMMIT_TYPE: {
      if (rest.length !== 1) {
        throw fail();
      }
      const [id] = rest;
      return {type: COMMIT_TYPE, id};
    }
    case USERLIKE_TYPE: {
      if (rest.length !== 2) {
        throw fail();
      }
      const [subtype, login] = rest;
      if (subtype !== "USER" && subtype !== "BOT") {
        throw fail();
      }
      return {type: USERLIKE_TYPE, subtype, login};
    }
    default:
      throw fail();
  }
}

export function toRaw(x: StructuredAddress): RawAddress {
  switch (x.type) {
    case REPO_TYPE:
      return NodeAddress.append(Prefix.repo, x.owner, x.name);
    case ISSUE_TYPE:
      return NodeAddress.append(
        Prefix.issue,
        x.repo.owner,
        x.repo.name,
        x.number
      );
    case PULL_TYPE:
      return NodeAddress.append(
        Prefix.pull,
        x.repo.owner,
        x.repo.name,
        x.number
      );
    case REVIEW_TYPE:
      return NodeAddress.append(
        Prefix.review,
        x.pull.repo.owner,
        x.pull.repo.name,
        x.pull.number,
        x.id
      );
    case COMMENT_TYPE:
      switch (x.parent.type) {
        case ISSUE_TYPE:
          return NodeAddress.append(
            Prefix.issueComment,
            x.parent.repo.owner,
            x.parent.repo.name,
            x.parent.number,
            x.id
          );
        case PULL_TYPE:
          return NodeAddress.append(
            Prefix.pullComment,
            x.parent.repo.owner,
            x.parent.repo.name,
            x.parent.number,
            x.id
          );
        case REVIEW_TYPE:
          return NodeAddress.append(
            Prefix.reviewComment,
            x.parent.pull.repo.owner,
            x.parent.pull.repo.name,
            x.parent.pull.number,
            x.parent.id,
            x.id
          );
        default:
          throw new Error(`Bad comment parent type: ${(x.parent.type: empty)}`);
      }
    case USERLIKE_TYPE:
      switch (x.subtype) {
        case "BOT":
          return NodeAddress.append(Prefix.bot, x.login);
        case "USER":
          return NodeAddress.append(Prefix.user, x.login);
        default:
          throw new Error((x.subtype: empty));
      }
    case COMMIT_TYPE:
      return NodeAddress.append(Prefix.commit, x.id);
    default:
      throw new Error(`Unexpected type ${(x.type: empty)}`);
  }
}
