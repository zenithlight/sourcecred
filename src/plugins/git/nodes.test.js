// @flow

import {NodeAddress} from "../../core/graph";
import * as GN from "./nodes";
import {fromRaw, toRaw} from "./nodes";

describe("plugins/git/nodes", () => {
  const examples = {
    commit: (): GN.CommitAddress => ({
      type: GN.COMMIT_TYPE,
      hash: "3715ddfb8d4c4fd2a6f6af75488c82f84c92ec2f",
    }),
  };

  // Incorrect types should be caught statically
  // $FlowExpectedError[prop-missing]
  const _unused_badTree: GN.RepoAddress = {
    type: "TREEEEE",
    hash: "browns",
  };

  describe("`fromRaw` after `toRaw` is identity", () => {
    Object.keys(examples).forEach((example) => {
      it(example, () => {
        const instance = examples[example]();
        expect(fromRaw(toRaw(instance))).toEqual(instance);
      });
    });
  });

  describe("`toRaw` after `fromRaw` is identity", () => {
    Object.keys(examples).forEach((example) => {
      it(example, () => {
        const instance = examples[example]();
        const raw = toRaw(instance);
        expect(toRaw(fromRaw(raw))).toEqual(raw);
      });
    });
  });

  describe("snapshots as expected:", () => {
    Object.keys(examples).forEach((example) => {
      it(example, () => {
        const instance = examples[example]();
        const raw = NodeAddress.toParts(toRaw(instance));
        expect({address: raw, structured: instance}).toMatchSnapshot();
      });
    });
  });

  describe("errors on", () => {
    describe("fromRaw(...) with", () => {
      function expectBadAddress(name: string, parts: $ReadOnlyArray<string>) {
        it(name, () => {
          const address = GN._gitAddress(...parts);
          expect(() => fromRaw(address)).toThrow("Bad address");
        });
      }
      it("undefined", () => {
        // $FlowExpectedError[incompatible-call]
        expect(() => fromRaw(undefined)).toThrow("undefined");
      });
      it("null", () => {
        // $FlowExpectedError[incompatible-call]
        expect(() => fromRaw(null)).toThrow("null");
      });
      it("bad prefix", () => {
        // $FlowExpectedError[incompatible-call]
        expect(() => fromRaw(NodeAddress.fromParts(["foo"]))).toThrow(
          "Bad address"
        );
      });

      expectBadAddress("no type", []);
      expectBadAddress("bad type", ["wat"]);

      expectBadAddress("commit with no hash", [GN.COMMIT_TYPE]);
      expectBadAddress("commit with extra field", [
        GN.COMMIT_TYPE,
        examples.commit().hash,
        examples.commit().hash,
      ]);
    });

    describe("toRaw(...) with", () => {
      it("null", () => {
        // $FlowExpectedError[incompatible-call]
        expect(() => toRaw(null)).toThrow("null");
      });
      it("undefined", () => {
        // $FlowExpectedError[incompatible-call]
        expect(() => toRaw(undefined)).toThrow("undefined");
      });
      it("bad type", () => {
        // $FlowExpectedError[incompatible-call]
        // $FlowExpectedError[prop-missing]
        expect(() => toRaw({type: "ICE_CREAM"})).toThrow("Unexpected type");
      });
    });
  });
});
