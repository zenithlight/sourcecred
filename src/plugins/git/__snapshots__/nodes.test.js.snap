// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`plugins/git/nodes snapshots as expected: commit 1`] = `
Object {
  "address": Array [
    "sourcecred",
    "git",
    "COMMIT",
    "3715ddfb8d4c4fd2a6f6af75488c82f84c92ec2f",
  ],
  "structured": Object {
    "hash": "3715ddfb8d4c4fd2a6f6af75488c82f84c92ec2f",
    "type": "COMMIT",
  },
}
`;
